### PR TITLE
Adding swagger v2 generator

### PR DESCRIPTION
**Pull request by gwendolyngoetz**
### Description

Adding schema generator to create a swagger definition for a program-specific the application api endpoint.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [x] Created unit and/or browser tests which fail without the change (if possible)

### Issue(s) this completes

Related to https://github.com/civiform/civiform/issues/5358

### CODE REVIEW

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+          containerDefinition.addDefinition(
+              Definition.builder(fieldName, definitionType)
+                  .setFormat(swaggerFormat)
+                  .setNullable(nullable)
+                  .build());
+        }
+      } else {
+        Definition enumeratorEntitiesDefinition =
+            Definition.builder("entities", DefinitionType.ARRAY)
+                .addDefinitions(buildApplicationDefinitions(childQuestionDefinitionNode))
```
Comment (avaleske) [inline server/app/services/openapi/v2/Swagger2SchemaGenerator.java:243]: there's also an `entity_name` string field in each entity

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+import services.openapi.v2.serializers.OpenApiSerializationAsserter;
+import services.program.BlockDefinition;
+import services.program.ProgramDefinition;
+import services.program.ProgramQuestionDefinition;
+import services.program.ProgramType;
+import services.program.StatusDefinitions;
+import services.question.types.QuestionDefinition;
+
+public class Swagger2SchemaGeneratorTest extends OpenApiSerializationAsserter {
+  private static final Stream<QuestionDefinition> ALL_SAMPLE_QUESTION_DEFINITIONS_WITH_IDS_STREAM =
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:24]: server/test/support/TestQuestionBank.java might be helpful here, as well as the FakeProgramBuilder utility in server/test/services/export/AbstractExporterTest.java

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("        type: array")
+            .appendLine("        items:")
+            .appendLine("          type: object")
+            .appendLine("          properties:")
+            .appendLine("            applicant_id:")
+            .appendLine("              type: integer")
+            .appendLine("              format: int32")
+            .appendLine("            application:")
+            .appendLine("              type: object")
+            .appendLine("              properties:")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:170]: **suggestion**: Would it make sense for each question type to be its own definition, that we can then re-use and compose into application definitions?

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                  type: object")
+            .appendLine("                  properties:")
+            .appendLine("                    question_type:")
+            .appendLine("                      type: string")
+            .appendLine("                    city:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    corrected:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:195]: **suggestion**: This is an enum, but I don't think we have an enum type yet in this serializer.

One thought around enums is that I've read that the default swagger generated consuming code can sometimes assume that the current enum values are all that will ever be present, rather then being built to handle future unknown values, so maybe leaving it as a string is better, to force a human to think about it?

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                    latitude:")
+            .appendLine("                      type: number")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:197]: **issue**: lat is unfortunately a string

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                      format: double")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    line2:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    longitude:")
+            .appendLine("                      type: number")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:204]: **issue**: long is also unfortunately a string

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    state:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    street:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    well_known_id:")
+            .appendLine("                      type: integer")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:217]: **issue**: well_known_id is also unfortunately a string.

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                    longitude:")
+            .appendLine("                      type: number")
+            .appendLine("                      format: double")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    service_area:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    state:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:212]: **thought**: This is effectively an enum, not sure if that's worth calling out.

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                sample_checkbox_question:")
+            .appendLine("                  type: object")
+            .appendLine("                  properties:")
+            .appendLine("                    question_type:")
+            .appendLine("                      type: string")
+            .appendLine("                    selections:")
+            .appendLine("                      type: array")
+            .appendLine("                      items:")
+            .appendLine("                        type: object")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:231]: **issue**: The items in the array are always strings

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("            applicant_id:")
+            .appendLine("              type: integer")
+            .appendLine("              format: int32")
+            .appendLine("            application:")
+            .appendLine("              type: object")
+            .appendLine("              properties:")
+            .appendLine("                name:")
+            .appendLine("                  type: object")
+            .appendLine("                  properties:")
+            .appendLine("                    question_type:")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:174]: **suggestion**: I discuss enums more below, but question_type is effectively an enum

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+            .appendLine("                        type: object")
+            .appendLine("                sample_file_upload_question:")
+            .appendLine("                  type: object")
+            .appendLine("                  properties:")
+            .appendLine("                    question_type:")
+            .appendLine("                      type: string")
+            .appendLine("                    file_key:")
+            .appendLine("                      type: string")
+            .appendLine("                      x-nullable: true")
+            .appendLine("                    original_file_name:")
```
Comment (avaleske) [inline server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java:283]: **issue**: We don't expose this in the API

**Review by avaleske — CHANGES_REQUESTED** [review]
Left some comments around fields that have changed since this was first drafted.

In particular there nested repeated question doesn't seem to be showing up in the output.

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+              .addSecurityRequirement(SecurityRequirement.builder(SecurityType.BASIC).build())
+              .addSecurityDefinition(SecurityDefinition.basicBuilder().build())
+              .setPaths(
+                  Paths.builder()
+                      .addPathItem(
+                          PathItem.builder("/applications")
+                              .addOperation(
+                                  Operation.builder(OperationType.GET)
+                                      .setOperationId("get_applications")
+                                      .setDescription("Get Applications")
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:60]: Good idea. I'll make the changes in here

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+/** Configuration used to generate a program specific swagger 2 schema */
+public class Swagger2SchemaGenerator implements OpenApiSchemaGenerator {
+  private final OpenApiSchemaSettings openApiSchemaSettings;
+
+  public Swagger2SchemaGenerator(OpenApiSchemaSettings openApiSchemaSettings) {
+    this.openApiSchemaSettings = checkNotNull(openApiSchemaSettings);
+  }
+
+  @Override
+  public String createSchema(ProgramDefinition programDefinition) {
```
Comment (avaleske) [inline server/app/services/openapi/v2/Swagger2SchemaGenerator.java:35]: oh also, @gwendolyngoetz, I don't know how you want to structure this but the API now returns every question that has ever been part of a program, not just the questions in the current program definition.

the ProgramService as a `getAllVersionsFullProgramDefinition()` method that could be helpful, and then here we pull out all the unique question paths:
https://github.com/civiform/civiform/blob/52339de8483f1d6b75637f3a6b820ff5fee933c3/server/app/services/export/JsonExporterService.java#L124-L135

In the exporter context, we do it based on the result of `ReadOnlyApplicantProgramService::getSummaryDataAllAnswers()` cause we'll need the answers anyway and it's an easy way to get all the nested paths for enumerators. Here we don't have an application, so we'll need to do something else.

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+                                      .setOperationId("get_applications")
+                                      .setDescription("Get Applications")
+                                      .setSummary("Export applications")
+                                      .addProduces(MimeType.Json)
+                                      .addTag("programs")
+                                      .addResponse(
+                                          Response.builder(HttpStatusCode.OK, "For valid requests.")
+                                              .setSchema("#/definitions/result")
+                                              .addHeader(
+                                                  Header.builder(DefinitionType.STRING, "x-next")
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:68]: That would be a very useful property to add, but we don't currently have it.

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+                                              .setDescription(
+                                                  "An opaque, alphanumeric identifier for a"
+                                                      + " specific page of results. When included"
+                                                      + " CiviForm will return a page of results"
+                                                      + " corresponding to the token.")
+                                              .build())
+                                      .build())
+                              .build())
+                      .build())
+              .addDefinition(
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:123]: Not a bad idea. Will do it when I have time in a separate PR.

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+                                          Definition.builder(
+                                                  "application",
+                                                  DefinitionType.OBJECT,
+                                                  buildApplicationDefinitions(programDefinition))
+                                              .build(),
+                                          Definition.builder(
+                                                  "application_id", DefinitionType.INTEGER)
+                                              .setFormat(Format.INT32)
+                                              .build(),
+                                          Definition.builder("create_time", DefinitionType.STRING)
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:144]: Yep, added that.

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java (hunk 8 above)
Comment (gwendolyngoetz) [reply server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java]: I'm considering including enums, but not in swagger 2. openapi 3.x have better support and 3.1 specifically aligns with the json schema spec.

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java (hunk 10 above)
Comment (gwendolyngoetz) [reply server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java]: see reply on other comment

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java (hunk 13 above)
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:36]: Going to leave it as it in order to include the enum values. There's likely to be more of them than other questions that might have been removed. Once #8860 is done we can revisit.

File: server/app/services/openapi/v2/Swagger2SchemaGenerator.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/services/openapi/v2/Swagger2SchemaGenerator.java:69]: created issue

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java (hunk 4 above)
Comment (gwendolyngoetz) [reply server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java]: See notes on it here https://github.com/civiform/civiform/pull/7051#discussion_r1876976400

File: server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java (hunk 3 above)
Comment (gwendolyngoetz) [reply server/test/services/openapi/v2/Swagger2SchemaGeneratorTest.java]: I'm not following.

File: server/build.sbt [hunk 17]
```diff
@@ -119,7 +119,15 @@ lazy val root = (project in file("."))
...
-      "ch.qos.logback" % "logback-classic" % "1.5.12"
+      "ch.qos.logback" % "logback-classic" % "1.5.12",
+
+      // Swagger 2.0 Dependencies
+      "io.swagger" % "swagger-core" % "1.6.14",
+      "io.swagger" % "swagger-parser" % "1.0.72",
+
+      // OpenAPI 3.x Dependencies
+      "io.swagger.core.v3" % "swagger-core" % "2.2.26",
+      "io.swagger.parser.v3" % "swagger-parser" % "2.1.24"
```
Comment (gwendolyngoetz) [inline server/build.sbt:130]: Searching over a year ago all the code I found wanted typed models. I'm not sure how I missed this, but I'm ripping out the custom code in favor of this. It's well supported by swagger.io.

**Pull request closed without merging**